### PR TITLE
More ValueOption in compiler: part 3 (and the last)

### LINKED ISSUE #16286: Replace active pattern with structs in the compiler code
Part of the upcoming perf effort: should have less allocs and speed things up a bit.

The current best RegEx seems to be this one: `(?:let|val)(\s+rec)?\s+\(\|[^|)]+\|_\|\)`

### PR DESCRIPTION
**Pull request by psfinaki**
Closes #16286

If my RegEx is correct, this should be it.

---

Some [benchmarks](https://github.com/dotnet/fsharp/pull/16822/commits/8b64ab9f1cf2acf124ec6f08d887bea22de1fada#diff-023336e328f4a1e8eb172c442afb3df0a15634aaa44b2a900145b2cda6e9730f) (for what was easy to bench):

Before:
| Method                           | Mean      | Error     | StdDev    | Gen0   | Allocated |
|--------------------------------- |----------:|----------:|----------:|-------:|----------:|
| TaskCancelled_TargetPattern      |  9.627 ns | 0.4100 ns | 1.2024 ns | 0.0004 |      24 B |
| TaskCancelled_OtherPattern       |  4.766 ns | 0.4129 ns | 1.2173 ns |      - |         - |
| StopProcessing_TargetPattern     | 11.123 ns | 0.6545 ns | 1.9297 ns | 0.0004 |      24 B |
| StopProcessing_OtherPattern      |  4.433 ns | 0.2699 ns | 0.7957 ns |      - |         - |
| ConstToILFieldInit_TargetPattern | 15.940 ns | 0.9367 ns | 2.7472 ns | 0.0008 |      48 B |
| ConstToILFieldInit_OtherPattern  |  2.418 ns | 0.0852 ns | 0.1621 ns |      - |         - |

After:
| Method                           | Mean      | Error     | StdDev    | Gen0   | Allocated |
|--------------------------------- |----------:|----------:|----------:|-------:|----------:|
| TaskCancelled_TargetPattern      | 0.0010 ns | 0.0037 ns | 0.0066 ns |      - |         - |
| TaskCancelled_OtherPattern       | 3.5823 ns | 0.2947 ns | 0.8688 ns |      - |         - |
| StopProcessing_TargetPattern     | 2.2672 ns | 0.1675 ns | 0.4938 ns |      - |         - |
| StopProcessing_OtherPattern      | 2.9171 ns | 0.2328 ns | 0.6863 ns |      - |         - |
| ConstToILFieldInit_TargetPattern | 9.8902 ns | 0.4454 ns | 1.2707 ns | 0.0004 |      24 B |
| ConstToILFieldInit_OtherPattern  | 3.0223 ns | 0.1553 ns | 0.4432 ns |      - |         - |

### CODE REVIEW

File: tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.SurfaceArea.netstandard20.debug.bsl [hunk 1]
```diff
@@ -4928,54 +4928,54 @@ FSharp.Compiler.Symbols.FSharpExpr: FSharp.Compiler.Text.Range get_Range()
 FSharp.Compiler.Symbols.FSharpExpr: Microsoft.FSharp.Collections.FSharpList`1[FSharp.Compiler.Symbols.FSharpExpr] ImmediateSubExpressions
 FSharp.Compiler.Symbols.FSharpExpr: Microsoft.FSharp.Collections.FSharpList`1[FSharp.Compiler.Symbols.FSharpExpr] get_ImmediateSubExpressions()
 FSharp.Compiler.Symbols.FSharpExpr: System.String ToString()
-FSharp.Compiler.Symbols.FSharpExprPatterns: Microsoft.FSharp.Core.FSharpOption`1[FSharp.Compiler.Symbols.FSharpExpr] |AddressOf|_|(FSharp.Compiler.Symbols.FSharpExpr)
```
Comment (vzarytovskii) [inline tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.SurfaceArea.netstandard20.debug.bsl:4931]: Can't change any of these APIs, they are public surface.

**Review by vzarytovskii — CHANGES_REQUESTED** [review]

File: tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.SurfaceArea.netstandard20.debug.bsl (hunk 1 above)
Comment (psfinaki) [reply tests/FSharp.Compiler.Service.Tests/FSharp.Compiler.Service.SurfaceArea.netstandard20.debug.bsl]: Yeah I had this feeling. I reverted these particular changes

File: tests/benchmarks/FCSBenchmarks/CompilerServiceBenchmarks/ValueOptionBenchmarks.fs [hunk 2]
```diff
@@ -0,0 +1,56 @@
+﻿namespace FSharp.Compiler.Benchmarks
```
Comment (T-Gro) [inline tests/benchmarks/FCSBenchmarks/CompilerServiceBenchmarks/ValueOptionBenchmarks.fs:1]: What value does it serve to the reader to keep this specific file around in a compiler repository?

I had the same feeling with most of other 'test plain F# usage'-benchmarks kept in this compiler repo, but this one raises my confusion the most.

Comment (psfinaki) [reply tests/benchmarks/FCSBenchmarks/CompilerServiceBenchmarks/ValueOptionBenchmarks.fs:1]: The benchmarks are here to provide at least some proof of the PR's value. And for reviewers to check that I am benchmarking things correctly.

As opposed to some other benchmarks in the repo, I don't expect these ones to be used in the near future, so I don't deem them necessary. OTOH I don't expect them to break.

If you wish, I can transform them to a GH gist.

Comment (T-Gro) [reply tests/benchmarks/FCSBenchmarks/CompilerServiceBenchmarks/ValueOptionBenchmarks.fs:1]: IMO it's fine just to delete them, neither repo nor gist needed for that sort of thing.

**Review by vzarytovskii — CHANGES_REQUESTED** [review]
We'll have to wait with this one for now, since now we're having some live issues with SO

File: tests/benchmarks/FCSBenchmarks/CompilerServiceBenchmarks/ValueOptionBenchmarks.fs (hunk 2 above)
Comment (psfinaki) [reply tests/benchmarks/FCSBenchmarks/CompilerServiceBenchmarks/ValueOptionBenchmarks.fs]: Removed

**Comment by vzarytovskii** [comment]
I've merged couple of fixes and workarounds for the SOs in CEs and in big nested expressions. We can probably merge this one (extra would be to test AnyCpu variant on those).

**Comment by abelbraaksma** [comment]
These performance improvements are quite impressive and more than I'd expect from such change. Nice!

**Review by T-Gro — APPROVED** [review]

**Comment by psfinaki** [comment]
I ran the current version of `fscAnyCpu.exe` against the code from [this](https://github.com/dotnet/fsharp/pull/17002) and [this](https://github.com/dotnet/fsharp/pull/17076) PRs, don't notice any regressions there.

**Review by abonie — APPROVED** [review]

**Review by vzarytovskii — APPROVED** [review]

**Pull request merged**